File: _Datacube_api_libraries_old/python_library/datacubepp/datacubepp/validators.py
from typing import List, Dict, Any
import re
from bson import ObjectId

class ValidationError(Exception):
    """
    Custom exception for validation errors.

    Attributes:
        message (str): Error message.
        field (str): Field name related to the error.
        code (str): Optional error code.
    """
    def __init__(self, message: str, field: str = "", code: str = ""):
        super().__init__(message)
        self.message = message
        self.field = field
        self.code = code

    def __str__(self):
        return f"{self.message} (field: {self.field}, code: {self.code})"

def validate_database_name(name: str) -> None:
    """
    Validates the name of the database.

    Args:
        name (str): The database name.

    Raises:
        ValidationError: If the database name is invalid.
    """
    if not isinstance(name, str) or not name.strip():
        raise ValidationError(
            "Database name must be a non-empty string.",
            field="database_name",
            code="DB_NAME_INVALID"
        )

    if not re.match(r"^[a-zA-Z0-9_]+$", name):
        raise ValidationError(
            "Database name can only contain alphanumeric characters and underscores.",
            field="database_name",
            code="DB_NAME_FORMAT_ERROR"
        )

def validate_collections(collections: List[Dict[str, Any]]) -> None:
    """
    Validates the structure of collections.

    Args:
        collections (List[Dict[str, Any]]): A list of collection definitions.

    Raises:
        ValidationError: If the collection definitions are invalid.
    """
    if not isinstance(collections, list):
        raise ValidationError(
            "Collections must be a list of dictionaries.",
            field="collections",
            code="COLLECTIONS_NOT_LIST"
        )

    for collection in collections:
        if not isinstance(collection, dict):
            raise ValidationError(
                "Each collection must be a dictionary.",
                field="collections",
                code="COLLECTION_NOT_DICT"
            )

        if "name" not in collection or not isinstance(collection["name"], str) or not collection["name"].strip():
            raise ValidationError(
                "Each collection must have a valid 'name'.",
                field="collection_name",
                code="COLLECTION_NAME_INVALID"
            )

        if not re.match(r"^[a-zA-Z0-9_]+$", collection["name"]):
            raise ValidationError(
                f"Collection name '{collection['name']}' is invalid. Only alphanumeric characters and underscores are allowed.",
                field="collection_name",
                code="COLLECTION_NAME_FORMAT_ERROR"
            )

        if "fields" not in collection or not isinstance(collection["fields"], list):
            raise ValidationError(
                f"Collection '{collection['name']}' must have a 'fields' key with a list of field definitions.",
                field="fields",
                code="FIELDS_MISSING_OR_INVALID"
            )

        for field in collection["fields"]:
            if not isinstance(field, dict):
                raise ValidationError(
                    f"Fields in collection '{collection['name']}' must be dictionaries.",
                    field="field_definition",
                    code="FIELD_NOT_DICT"
                )

            if "name" not in field or not isinstance(field["name"], str) or not field["name"].strip():
                raise ValidationError(
                    f"Each field in collection '{collection['name']}' must have a valid 'name'.",
                    field="field_name",
                    code="FIELD_NAME_INVALID"
                )

            if not re.match(r"^[a-zA-Z0-9_]+$", field["name"]):
                raise ValidationError(
                    f"Field name '{field['name']}' in collection '{collection['name']}' is invalid.",
                    field="field_name",
                    code="FIELD_NAME_FORMAT_ERROR"
                )

            if "type" not in field or field["type"] not in {"string", "integer", "boolean", "float", "date"}:
                raise ValidationError(
                    f"Field '{field['name']}' in collection '{collection['name']}' has an invalid or missing type.",
                    field="field_type",
                    code="FIELD_TYPE_INVALID"
                )

def validate_object_id(object_id: str) -> None:
    """
    Validates if a string is a valid MongoDB ObjectId.

    Args:
        object_id (str): The ObjectId to validate.

    Raises:
        ValidationError: If the ObjectId is invalid.
    """
    if not isinstance(object_id, str) or not object_id.strip():
        raise ValidationError(
            "ObjectId must be a non-empty string.",
            field="object_id",
            code="OBJECT_ID_INVALID"
        )

    try:
        ObjectId(object_id)
    except Exception:
        raise ValidationError(
            "Invalid ObjectId format.",
            field="object_id",
            code="OBJECT_ID_FORMAT_ERROR"
        )

File: _Datacube_api_libraries_old/node_library/datacubenp/src/validators.ts
// validators.ts

import { ValidationError } from "./exceptions";

/**
 * Validates if a string is a valid MongoDB ObjectId.
 * @param id - The string to validate.
 */
export function validateDatabaseId(id: string): void {
  if (!id || typeof id !== "string" || !/^[a-fA-F0-9]{24}$/.test(id)) {
    throw new ValidationError(
      "Invalid database ID. Must be a 24-character hexadecimal string.",
    );
  }
}

/**
 * Validates if a collection name is a non-empty string.
 * @param name - The collection name to validate.
 */
export function validateCollectionName(name: string): void {
  if (!name || typeof name !== "string" || !/^[\w-]+$/.test(name)) {
    throw new ValidationError(
      "Collection name must be a non-empty string containing only alphanumeric characters, underscores, or hyphens.",
    );
  }
}

/**
 * Validates the structure of collections.
 * @param collections - The collections to validate.
 */
export function validateCollections(
  collections: Array<{
    name: string;
    fields: Array<{ name: string; type?: string }>;
  }>,
): void {
  if (!Array.isArray(collections)) {
    throw new ValidationError("Collections must be an array.");
  }

  collections.forEach((collection) => {
    if (!collection.name || typeof collection.name !== "string") {
      throw new ValidationError(
        'Each collection must have a "name" property as a non-empty string.',
      );
    }

    if (!Array.isArray(collection.fields)) {
      throw new ValidationError(
        `The "fields" property in collection "${collection.name}" must be an array.`,
      );
    }

    collection.fields.forEach((field) => {
      if (!field.name || typeof field.name !== "string") {
        throw new ValidationError(
          `Each field in collection "${collection.name}" must have a "name" property as a non-empty string.`,
        );
      }

      if (field.type && typeof field.type !== "string") {
        throw new ValidationError(
          `Field "${field.name}" in collection "${collection.name}" must have a "type" property as a string.`,
        );
      }
    });
  });
}

/**
 * Validates the structure of a document.
 * @param document - The document to validate.
 */
export function validateDocument(document: object): void {
  if (!document || typeof document !== "object" || Array.isArray(document)) {
    throw new ValidationError("Document must be a non-array object.");
  }
}
